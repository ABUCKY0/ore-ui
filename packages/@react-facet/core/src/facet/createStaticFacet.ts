import { Facet } from '../types'

export const createStaticFacet = <T>(value: T): Facet<T> => {
  return {
    get: () => value,
    observe: (listener) => {
      listener(value)
      return () => {}
    },
  }
}
