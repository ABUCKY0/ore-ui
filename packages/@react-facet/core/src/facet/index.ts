export * from './createFacet'
export * from './createReadOnlyFacet'
export * from './createStaticFacet'
export * from './wrapFacets'
