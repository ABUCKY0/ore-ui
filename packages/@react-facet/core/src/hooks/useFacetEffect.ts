import { ReactFacetDevTools } from '@react-facet/dev-tools'
import { useCallback, useLayoutEffect } from 'react'
import { Facet, Effect } from '../types'

/**
 * Allows using a selector to perform imperative code (manual DOM manipulations)
 *
 * @param effect function that will do the side-effect (ex: update the DOM)
 * @param dependencies variable used by the map that are available in scope (similar as dependencies of useEffect)
 * @param facet facet that the effect listens to
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 */
export function useFacetEffect<V>(
  effect: Effect<V>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dependencies: any[],
  facet: Facet<V>,
) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;((global as any).__REACT_FACET_DEVTOOLS_GLOBAL_HOOK__ as ReactFacetDevTools).send({
      hookName: 'useFacetEffect',
      facets: [facet],
    })
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const effectMemoized = useCallback(effect, dependencies)
  useLayoutEffect(() => facet.observe(effectMemoized), [facet, effectMemoized])
}
