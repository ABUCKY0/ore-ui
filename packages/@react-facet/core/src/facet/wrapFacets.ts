import { FacetProp, Facet, isFacet } from '../types'
import { createStaticFacet } from './createStaticFacet'

export const wrapFacets = (facetProps: FacetProp<unknown>[]): Facet<unknown>[] => {
  return facetProps.map((facetProp) => (isFacet(facetProp) ? facetProp : createStaticFacet(facetProp)))
}
