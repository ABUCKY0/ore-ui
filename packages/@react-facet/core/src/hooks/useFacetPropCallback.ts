import { useCallback, useLayoutEffect, useRef } from 'react'
import { wrapFacets } from '../facet'
import { NO_VALUE, NoValue, Option, FacetProp, ExtractFacetPropValues } from '../types'

/**
 * Creates a callback that depends on the value of a facet.
 * Very similar to `useCallback` from `React`
 *
 * @param callback function callback that receives the current facet values and the arguments passed to the callback
 * @param facetProps facets that the callback listens to
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 */
export function useFacetCallback<M, Y extends FacetProp<unknown>[], T extends [...Y], K extends [...unknown[]]>(
  callback: (...args: ExtractFacetPropValues<T>) => (...args: K) => M,
  facetProps: T,
): (...args: K) => M | NoValue {
  const facetsRef = useRef<Option<unknown>[]>(facetProps.map(() => NO_VALUE))

  useLayoutEffect(() => {
    const unsubscribes = wrapFacets(facetProps).map((facet, index) => {
      return facet.observe((value) => {
        facetsRef.current[index] = value
      })
    })

    return () => {
      unsubscribes.forEach((unsubscribe) => unsubscribe())
    }
    // We care about each individual facet and if any is a different reference
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, facetProps)

  // eslint-disable-next-line react-hooks/exhaustive-deps
  return useCallback(
    (...args: K) => {
      const values = facetsRef.current

      for (const value of values) {
        if (value === NO_VALUE) return NO_VALUE
      }

      return callback(...(values as ExtractFacetPropValues<T>))(...(args as K))
    },
    [callback, facetsRef],
  )
}
