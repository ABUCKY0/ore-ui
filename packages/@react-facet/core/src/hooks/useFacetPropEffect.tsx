import { useEffect } from 'react'
import { wrapFacets } from '../facet'
import { Unsubscribe, Cleanup, NO_VALUE, FacetProp, ExtractFacetPropValues, isFacet } from '../types'

/**
 * Allows running an effect based on facet updates. Similar to React's useEffect.
 *
 * @param effect function that will do the side-effect (ex: update the DOM)
 * @param facetProps facets that the effect listens to
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 */
export function useFacetPropEffect<Y extends FacetProp<unknown>[], T extends [...Y]>(
  effect: (...args: ExtractFacetPropValues<T>) => void | Cleanup,
  facetProps: T,
) {
  useEffect(() => {
    const effectTyped = effect as (...args: unknown[]) => ReturnType<typeof effect>

    if (facetProps.length === 1) {
      const facetProp = facetProps[0]

      return isFacet(facetProp) ? facetProp.observe(effectTyped) : effectTyped(facetProp)
    }

    let cleanup: void | Cleanup
    let hasAllDependencies = false
    const unsubscribes: Unsubscribe[] = []
    const values: unknown[] = facetProps.map(() => NO_VALUE)

    wrapFacets(facetProps).forEach((facet, index) => {
      unsubscribes[index] = facet.observe((value) => {
        values[index] = value
        hasAllDependencies = hasAllDependencies || values.every((value) => value != NO_VALUE)

        if (hasAllDependencies) {
          if (cleanup != null) {
            cleanup()
          }

          cleanup = effect(...(values as ExtractFacetPropValues<T>))
        }
      })
    })

    return () => {
      unsubscribes.forEach((unsubscribe) => unsubscribe())
      if (cleanup != null) {
        cleanup()
      }
    }

    // We care about each individual facet and if any is a different reference
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [effect, ...facetProps])
}
