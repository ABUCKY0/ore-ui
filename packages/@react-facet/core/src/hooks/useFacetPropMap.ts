import { useMemo } from 'react'
import { wrapFacets } from '../facet'
import { defaultEqualityCheck } from '../equalityChecks'
import { mapFacetsLightweight } from '../mapFacets'
import { EqualityCheck, Facet, FacetProp, NoValue, Value, ExtractFacetPropValues } from '../types'

/**
 * Helper hook that allows mapping a value from a facet with local variables/props in a React component
 *
 * @param selector function that takes value from provided facets and maps them to a new value
 * @param facetProps
 * @param equalityCheck optional, has a default for immutable values
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 *
 * @returns a new facet definition that can be consumed as a regular facet
 */
export function useFacetPropMap<M extends Value, Y extends FacetProp<unknown>[], T extends [...Y]>(
  selector: (...args: ExtractFacetPropValues<T>) => M | NoValue,
  facetProps: T,
  equalityCheck: EqualityCheck<M> = defaultEqualityCheck,
): Facet<M> {
  const facetComposition = useMemo<Facet<M>>(() => {
    const selectorTyped = selector as (...args: unknown[]) => ReturnType<typeof selector>

    return mapFacetsLightweight(wrapFacets(facetProps), selectorTyped, equalityCheck)

    // We need to disable the linter on the next line given we are spreading the facets as individual dependencies
    // of the effect. We do this to avoid re-running this effect when passing a new array with the same facets.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selector, equalityCheck, ...facetProps])

  return facetComposition
}
