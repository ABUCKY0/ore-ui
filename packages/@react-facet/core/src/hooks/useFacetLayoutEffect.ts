import { useLayoutEffect } from 'react'
import { createUseFacetEffect } from './useFacetEffect'

/**
 * Allows running an effect based on facet updates. Similar to React's useLayoutEffect.
 *
 * @param effect function that will do the side-effect (ex: update the DOM)
 * @param dependencies variable used by the map that are available in scope (similar as dependencies of useEffect)
 * @param facets facets that the effect listens to
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 */
export const useFacetLayoutEffect = createUseFacetEffect(useLayoutEffect)
