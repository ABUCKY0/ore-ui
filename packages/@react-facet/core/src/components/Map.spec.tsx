import React, { useEffect } from 'react'
import { act, render } from '@react-facet/dom-fiber-testing-library'
import { createFacet } from '../facet'
import { Facet, NoValue } from '../types'
import { useFacetEffect, useFacetMap } from '../hooks'
import { Map, MapWithKey } from '.'

describe('<Map />', () => {
  it('renders all items in a Facet of array', () => {
    type Input = { a: string }
    const data = createFacet({
      initialValue: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }, { a: '5' }],
    })

    const ExampleContent = ({ item, index }: { item: Facet<Input>; index: number }) => {
      return (
        <span>
          <fast-text text={useFacetMap(({ a }) => a, [], [item])} />
          <span>{index}</span>
        </span>
      )
    }

    const inputEqualityCheck = () => {
      const previous: Partial<Input> = {}

      return (current: Input) => {
        if (current.a === previous.a) {
          return true
        }
        previous.a = current.a
        return false
      }
    }
    const Example = () => {
      return (
        <Map array={data} equalityCheck={inputEqualityCheck}>
          {(item, index) => <ExampleContent item={item} index={index} />}
        </Map>
      )
    }

    const scenario = <Example />

    const { container } = render(scenario)

    expect(container).toMatchSnapshot()
  })

  it('unmounts components when the array reduces in size', () => {
    interface Item {
      value: string
    }

    const data = createFacet<Item[]>({
      initialValue: [{ value: '1' }, { value: '2' }, { value: '3' }, { value: '4' }, { value: '5' }],
    })

    const Item = ({ item, index }: { item: Facet<Item>; index: number }) => (
      <span>
        <fast-text text={useFacetMap(({ value }) => value, [], [item])} />
        <span>{index}</span>
      </span>
    )

    const itemEqualityCheck = () => {
      const previous: Partial<Item> = {}

      return (current: Item) => {
        if (current.value === previous.value) {
          return true
        }

        previous.value = current.value

        return false
      }
    }

    const Example = () => {
      return (
        <Map array={data} equalityCheck={itemEqualityCheck}>
          {(item, index) => <Item item={item} index={index} />}
        </Map>
      )
    }

    const scenario = <Example />

    const { container } = render(scenario)

    expect(container).toMatchSnapshot()

    data.set([{ value: '1' }, { value: '2' }])

    expect(container).toMatchSnapshot()
  })

  it('updates only items that have changed', () => {
    type Input = { a: string }
    const data = createFacet({
      initialValue: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }, { a: '5' }],
    })

    const mock = jest.fn()

    const ExampleContent = ({ item }: { item: Facet<Input> }) => {
      useFacetEffect(mock, [], [item])
      return null
    }

    const inputEqualityCheck = () => {
      const previous: Partial<Input> = {}

      return (current: Input) => {
        if (current.a === previous.a) {
          return true
        }

        previous.a = current.a

        return false
      }
    }

    const Example = () => {
      return (
        <Map array={data} equalityCheck={inputEqualityCheck}>
          {(item) => <ExampleContent item={item} />}
        </Map>
      )
    }

    const scenario = <Example />

    render(scenario)

    expect(mock).toHaveBeenCalledTimes(5)

    mock.mockClear()

    act(() => {
      data.set([{ a: '6' }, { a: '2' }, { a: '3' }, { a: '4' }, { a: '5' }])
    })

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith({ a: '6' })
  })
})

describe('<MapWithKey />', () => {
  it('renders all items in a Facet of array', () => {
    type Input = { key: string; data: string }
    const data = createFacet<Input[]>({
      initialValue: [
        { key: '1', data: 'a' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
      ],
    })

    const ExampleContent = ({ item, index }: { item: Facet<Input>; index: number }) => {
      return (
        <span>
          <fast-text text={useFacetMap(({ data }) => data, [], [item])} />
          <span>{index}</span>
        </span>
      )
    }

    const keySelector = ({ key }: Input) => key

    const inputEqualityCheck = () => {
      const previous: Partial<Input> = {}

      return (current: Input) => {
        if (current.key === previous.key && current.data === previous.data) {
          return true
        }

        previous.key = current.key
        previous.data = current.data

        return false
      }
    }

    const Example = () => {
      return (
        <MapWithKey array={data} equalityCheck={inputEqualityCheck} keySelector={keySelector}>
          {(item, index) => <ExampleContent item={item} index={index} />}
        </MapWithKey>
      )
    }

    const scenario = <Example />

    const { container } = render(scenario)

    expect(container).toMatchSnapshot()
  })

  it('unmounts components when the array reduces in size', () => {
    type Input = { key: string; data: string }
    const data = createFacet<Input[]>({
      initialValue: [
        { key: '1', data: 'a' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
      ],
    })

    const Item = ({ item, index }: { item: Facet<Input>; index: number }) => (
      <span>
        <fast-text text={useFacetMap(({ data }) => data, [], [item])} />
        <span>{index}</span>
      </span>
    )

    const keySelector = ({ key }: Input) => key

    const itemEqualityCheck = () => {
      const previous: Partial<Input> = {}

      return (current: Input) => {
        if (current.key === previous.key && current.data === previous.data) {
          return true
        }

        previous.key = current.key
        previous.data = current.data

        return false
      }
    }

    const Example = () => {
      return (
        <MapWithKey array={data} equalityCheck={itemEqualityCheck} keySelector={keySelector}>
          {(item, index) => <Item item={item} index={index} />}
        </MapWithKey>
      )
    }

    const scenario = <Example />

    const { container } = render(scenario)

    expect(container).toMatchSnapshot()

    data.set([
      { key: '1', data: 'a' },
      { key: '2', data: 'b' },
    ])

    expect(container).toMatchSnapshot()
  })

  it('updates only items that have changed', () => {
    type Input = { key: string; data: string }
    const data = createFacet<Input[]>({
      initialValue: [
        { key: '1', data: 'a' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
      ],
    })

    const mock = jest.fn()

    const ExampleContent = ({ item }: { item: Facet<Input> }) => {
      useFacetEffect(mock, [], [item])
      return null
    }

    const keySelector = ({ key }: Input) => key

    const inputEqualityCheck = () => {
      const previous: Partial<Input> = {}

      return (current: Input) => {
        if (current.key === previous.key && current.data === previous.data) {
          return true
        }

        previous.key = current.key
        previous.data = current.data

        return false
      }
    }

    const Example = () => {
      return (
        <MapWithKey array={data} equalityCheck={inputEqualityCheck} keySelector={keySelector}>
          {(item) => <ExampleContent item={item} />}
        </MapWithKey>
      )
    }

    const scenario = <Example />

    render(scenario)

    expect(mock).toHaveBeenCalledTimes(5)

    mock.mockClear()

    act(() => {
      data.set([
        { key: '1', data: 'z' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
      ])
    })

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith({ key: '1', data: 'z' })
  })

  it('rerenders only when key is updated or added', () => {
    type Input = { key: string; data: string }
    const data = createFacet<Input[]>({
      initialValue: [
        { key: '1', data: 'a' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
      ],
    })

    const facetUpdateMock = jest.fn()
    const onMount = jest.fn()

    const ExampleContent = ({
      item,
      index,
    }: {
      index: number
      item: Facet<Input>
      onMount: (input: Input | NoValue) => void
    }) => {
      // eslint-disable-next-line react-hooks/exhaustive-deps
      useEffect(() => onMount(index), [])
      useFacetEffect((itemValue) => facetUpdateMock(itemValue, index), [index], [item])
      const dataFacet = useFacetMap(({ data }) => `Data: ${data}`, [], [item])
      const keyFacet = useFacetMap(({ key }) => `Key: ${key}`, [], [item])

      return (
        <div>
          <p>
            <fast-text text={keyFacet} />
          </p>
          <p>
            <fast-text text={dataFacet} />
          </p>
        </div>
      )
    }

    const keySelector = ({ key }: Input) => key

    const inputEqualityCheck = () => {
      const previous: Partial<Input> = {}

      return (current: Input) => {
        if (current.key === previous.key && current.data === previous.data) {
          return true
        }

        previous.key = current.key
        previous.data = current.data

        return false
      }
    }

    const Example = () => {
      return (
        <MapWithKey array={data} equalityCheck={inputEqualityCheck} keySelector={keySelector}>
          {(item, index) => <ExampleContent item={item} index={index} onMount={facetUpdateMock} />}
        </MapWithKey>
      )
    }

    const scenario = <Example />

    const { container } = render(scenario)

    expect(facetUpdateMock).toHaveBeenCalledTimes(5)
    expect(onMount).toHaveBeenCalledTimes(5)

    expect(container).toMatchSnapshot()

    facetUpdateMock.mockClear()
    onMount.mockClear()

    act(() => {
      data.set([
        { key: '1', data: 'z' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
      ])
    })

    expect(facetUpdateMock).toHaveBeenCalledTimes(1)
    expect(facetUpdateMock).toHaveBeenCalledWith({ key: '1', data: 'z' }, 0)
    expect(onMount).not.toHaveBeenCalled()

    expect(container).toMatchSnapshot()

    facetUpdateMock.mockClear()
    onMount.mockClear()

    act(() => {
      data.set([
        { key: '1', data: 'z' },
        { key: '2', data: 'b' },
        { key: '3', data: 'c' },
        { key: '4', data: 'd' },
        { key: '5', data: 'e' },
        { key: '6', data: 'f' },
      ])
    })

    expect(facetUpdateMock).toHaveBeenCalledTimes(1)
    expect(facetUpdateMock).toHaveBeenCalledWith({ key: '6', data: 'f' }, 5)
    expect(onMount).toHaveBeenCalledWith(5)

    expect(container).toMatchSnapshot()

    facetUpdateMock.mockClear()
    onMount.mockClear()
  })
})
