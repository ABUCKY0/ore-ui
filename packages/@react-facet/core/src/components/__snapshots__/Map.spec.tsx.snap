// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> renders all items in a Facet of array 1`] = `
<div>
  <span>
    1
    <span>
      0
    </span>
  </span>
  <span>
    2
    <span>
      1
    </span>
  </span>
  <span>
    3
    <span>
      2
    </span>
  </span>
  <span>
    4
    <span>
      3
    </span>
  </span>
  <span>
    5
    <span>
      4
    </span>
  </span>
</div>
`;

exports[`<Map /> unmounts components when the array reduces in size 1`] = `
<div>
  <span>
    1
    <span>
      0
    </span>
  </span>
  <span>
    2
    <span>
      1
    </span>
  </span>
  <span>
    3
    <span>
      2
    </span>
  </span>
  <span>
    4
    <span>
      3
    </span>
  </span>
  <span>
    5
    <span>
      4
    </span>
  </span>
</div>
`;

exports[`<Map /> unmounts components when the array reduces in size 2`] = `
<div>
  <span>
    1
    <span>
      0
    </span>
  </span>
  <span>
    2
    <span>
      1
    </span>
  </span>
</div>
`;

exports[`<MapWithKey /> renders all items in a Facet of array 1`] = `
<div>
  <span>
    a
    <span>
      0
    </span>
  </span>
  <span>
    b
    <span>
      1
    </span>
  </span>
  <span>
    c
    <span>
      2
    </span>
  </span>
  <span>
    d
    <span>
      3
    </span>
  </span>
  <span>
    e
    <span>
      4
    </span>
  </span>
</div>
`;

exports[`<MapWithKey /> rerenders only when key is updated or added 1`] = `
<div>
  <div>
    <p>
      Key: 1
    </p>
    <p>
      Data: a
    </p>
  </div>
  <div>
    <p>
      Key: 2
    </p>
    <p>
      Data: b
    </p>
  </div>
  <div>
    <p>
      Key: 3
    </p>
    <p>
      Data: c
    </p>
  </div>
  <div>
    <p>
      Key: 4
    </p>
    <p>
      Data: d
    </p>
  </div>
  <div>
    <p>
      Key: 5
    </p>
    <p>
      Data: e
    </p>
  </div>
</div>
`;

exports[`<MapWithKey /> rerenders only when key is updated or added 2`] = `
<div>
  <div>
    <p>
      Key: 1
    </p>
    <p>
      Data: z
    </p>
  </div>
  <div>
    <p>
      Key: 2
    </p>
    <p>
      Data: b
    </p>
  </div>
  <div>
    <p>
      Key: 3
    </p>
    <p>
      Data: c
    </p>
  </div>
  <div>
    <p>
      Key: 4
    </p>
    <p>
      Data: d
    </p>
  </div>
  <div>
    <p>
      Key: 5
    </p>
    <p>
      Data: e
    </p>
  </div>
</div>
`;

exports[`<MapWithKey /> rerenders only when key is updated or added 3`] = `
<div>
  <div>
    <p>
      Key: 1
    </p>
    <p>
      Data: z
    </p>
  </div>
  <div>
    <p>
      Key: 2
    </p>
    <p>
      Data: b
    </p>
  </div>
  <div>
    <p>
      Key: 3
    </p>
    <p>
      Data: c
    </p>
  </div>
  <div>
    <p>
      Key: 4
    </p>
    <p>
      Data: d
    </p>
  </div>
  <div>
    <p>
      Key: 5
    </p>
    <p>
      Data: e
    </p>
  </div>
  <div>
    <p>
      Key: 6
    </p>
    <p>
      Data: f
    </p>
  </div>
</div>
`;

exports[`<MapWithKey /> unmounts components when the array reduces in size 1`] = `
<div>
  <span>
    a
    <span>
      0
    </span>
  </span>
  <span>
    b
    <span>
      1
    </span>
  </span>
  <span>
    c
    <span>
      2
    </span>
  </span>
  <span>
    d
    <span>
      3
    </span>
  </span>
  <span>
    e
    <span>
      4
    </span>
  </span>
</div>
`;

exports[`<MapWithKey /> unmounts components when the array reduces in size 2`] = `
<div>
  <span>
    a
    <span>
      0
    </span>
  </span>
  <span>
    b
    <span>
      1
    </span>
  </span>
</div>
`;
